x-pack/test/observability_ai_assistant_functional/tests/contextual_insights/index.spec.ts

```diff
@@ -0,0 +1,168 @@
+            .failure()
+        ),
+    ];
+
+    await apmSynthtraceEsClient.index(documents);
+  }
+
+  async function createOpenAiConnector(proxy: LlmProxy) {
+    await supertest
+      .post('/api/actions/connector')
+      .set('kbn-xsrf', 'foo')
+      .send({
+        name: 'foo',
+        config: {
+          apiProvider: 'OpenAI',
+          apiUrl: `http://localhost:${proxy.getPort()}`,
+          defaultModel: 'gpt-4',
+        },
+        secrets: { apiKey: 'myApiKey' },
+        connector_type_id: '.gen-ai',
+      })
+      .expect(200);
+  }
+
+  async function deleteConnectors() {
+    const connectors = await supertest.get('/api/actions/connectors').expect(200);
+    const promises = connectors.body.map((connector: { id: string }) => {
+      return supertest
+        .delete(`/api/actions/connector/${connector.id}`)
+        .set('kbn-xsrf', 'foo')
+        .expect(204);
+    });
+
+    try {
+      await Promise.all(promises);
+    } catch (e) {
+      if (e?.response?.status === 404) {
+        return;
+      }
+
+      log.debug('Error deleting connector');
+      throw e;
+    }
+  }
+
+  async function navigateToError() {
+    await common.navigateToApp('apm', {
+      hash: `/services/opbeans-go/errors?rangeFrom=now-15m&rangeTo=now`,
+    });
+
+    await testSubjects.click('errorGroupId');
+  }
+
+  describe('Contextual insights for APM errors', () => {
+    let proxy: LlmProxy;
+
+    before(async () => {
+      await Promise.all([
+        deleteConnectors(), // cleanup previous connectors
+        apmSynthtraceEsClient.clean(), // cleanup previous synthtrace data
+      ]);
+
+      await Promise.all([
+        createSynthtraceErrors(), // synthtrace data
+        ui.auth.login(), // login
+      ]);
+    });
+
+    describe('when there are no connectors', () => {
+      it('should not show the contextual insight component', async () => {
+        await navigateToError();
+        await testSubjects.missingOrFail(ui.pages.contextualInsights.button);
+      });
+    });
+
+    describe('when there are connectors', () => {
+      before(async () => {
+        proxy = await createLlmProxy(log);
+        await createOpenAiConnector(proxy);
+      });
+
+      it('should show the contextual insight component on the APM error details page', async () => {
+        await navigateToError();
+
+        const waitForIntercept = proxy
+          .intercept(
+            'conversation',
+            (body) => !isFunctionTitleRequest(body),
+            'This error is nothing to worry about. Have a nice day!'
+          )
+          .complete();
+
+        await testSubjects.click(ui.pages.contextualInsights.button);
+
+        await retry.try(async () => {
+          const llmResponse = await testSubjects.getVisibleText(ui.pages.contextualInsights.text);
+          expect(llmResponse).to.contain('This error is nothing to worry about. Have a nice day!');
+        });
+
+        await waitForIntercept;
```

Comment:
I see `await waitForIntercept;` as an assertion in itself. I would prefer something like what we have for jest though:
```
expect(intercept).toHaveBeenCalled()
```

In this case it doesn't matter if it is before or after the assertion since we are doing retries until the response is available (which is almost instant). 

Happy to move around though

Reply:
It's not `waitForIntercept` here - you're passing chunks, so it will automatically complete for you. What you are awaiting is the response from the LLM to be finished.

Reply:
True, it waits for both the intercept and for the response to finish:

https://github.com/elastic/kibana/blob/922900654288002f120f65f32f0d50f27d0ab1cb/x-pack/test/observability_ai_assistant_api_integration/common/create_llm_proxy.ts#L166-L170

Still, is that important to wait for in this test? Shouldn't we test the UI, and not implementation details?

Reply:
I'm not sure if I understand why we are testing implementation details. You're essentially mocking a response here, and you want to wait until that mocked response has been processed by the UI before you run your assertion. Isn't this the reason why the tests are flaky?